### PR TITLE
registry<>: add Payload template parameter to separate identity from value

### LINKED ISSUE #400: registry<T, Tag>: separate identity from payload (custom Hash/Eq or <Key, Payload> form)
## Problem

`registry<T, Tag>` deduplicates on insert by looking values up in
`unordered_map<T, id_type>`, which uses `T`'s natural `operator==` and
`std::hash<T>`. That's the right behavior when identity *is* the whole
value (`std::string` intern pools, integer pools, etc.).

It becomes awkward when the registry stores a struct where **identity is a
subset** of the fields. Example from a downstream consumer (atroplex):

```cpp
struct gene_info {
    std::string gene_id;       // identity
    std::string gene_name;     // payload (canonical name)
    std::string gene_biotype;  // payload (canonical biotype)
};
```

The consumer wants:

1. Lookup by `gene_id` alone (two `gene_info` values with the same id collapse
   to one slot).
2. First-write-wins on payload (annotations sort first; later TALON entries
   with the same `gene_id` but placeholder name/biotype must reuse the
   existing id and **silently drop** their placeholder fields).

Today the only way to get that from `registry<T, Tag>` is to overload
`gene_info::operator==` and `std::hash<gene_info>` to consider `gene_id` only.
That works — `intern()` returns the existing id on a hit, payload is
preserved — but introduces a **partial-equality footgun**: any consumer
holding `gene_info` values gets surprising `a == b` behavior anywhere outside
the registry (set membership, sort dedup, equality assertions in tests).

### Option A — custom Hash/Eq template parameters

```cpp
template<registry_value T,
         typename Tag = void,
         typename Hash = std::hash<T>,
         typename Eq = std::equal_to<T>>
class registry { ... };
```

Lets consumers pass id-only hash/eq without overloading `T`'s natural
operators:

```cpp
struct gene_id_hash { size_t operator()(const gene_info& g) const noexcept {
    return std::hash<std::string>{}(g.gene_id); } };
struct gene_id_eq   { bool   operator()(const gene_info& a, const gene_info& b) const {
    return a.gene_id == b.gene_id; } };

using gene_registry = registry<gene_info, gene_tag, gene_id_hash, gene_id_eq>;
```

`gene_info`'s natural `operator==` stays deep / honest. Smallest change to
the existing API.

### Option B — `<Key, Payload, Tag>` form

```cpp
template<registry_value Key,
         typename Payload = Key,
         typename Tag = void>
class registry {
    // intern(key, payload) for Key != Payload; intern(value) when same
    id_type intern(const Key& key, const Payload& payload);
    const Payload& get(id_type id) const;
};
```

Cleaner end state — separation of identity and payload is explicit in the
type. Larger API change.

## Why this matters beyond gene_info

Any registry of struct values where identity is a subset of fields hits the
same issue: sample metadata keyed on sample_id, file records keyed on path,
etc. Today the workaround (overload `operator==`) leaks partial equality to
every consumer that holds the type.

## Workaround in use today

atroplex is shipping the bare alias + partial-equality `operator==`
workaround in the meantime (with a load-bearing comment), and will migrate
to whichever form lands.

## Suggested

Option A is the smaller, fully-backwards-compatible change. Existing
`registry<T, Tag>` users see no diff (defaults). Option B is cleaner long
term but is the bigger surface change.

### PR DESCRIPTION
**Pull request by riasc**
### Added

- Adds a third template parameter `Payload` (default `= Key`) so `registry<>` can store records keyed on a subset of their own fields, addressing the gene_info / partial-equality footgun in #400.
- New `intern(key, payload)` overload deduplicates on `key` alone with **first-write-wins** semantics — re-interning the same key with a later payload returns the existing id and silently drops the new payload.
- Backwards compatible: original `intern(value)` form is kept via `requires Key == Payload`. Existing `registry<T>` and `registry<T, Tag>` call sites compile and behave identically. Wire format is unchanged when `Key == Payload`.

Closes #400.

## Design notes

Template ordering is `<Key, Tag = void, Payload = Key>` (Tag in the middle) rather than the `<Key, Payload, Tag>` ordering sketched in the issue. This preserves the existing public surface — `registry<std::string, transcript_tag>` still means what it used to. Reaching the new functionality looks like `registry<std::string, void, gene_info>` or, more naturally, `registry<std::string, gene_tag, gene_info>`.

Serialization stays on the historical `count + payloads` layout when `Key == Payload` (no on-disk migration needed) and switches to `count + (key, payload) pairs` only when the types differ. `deserialize` keeps the strong exception guarantee in both branches.

## Test plan

- [x] I, as a human being, have checked each line of code
- [x] All previously-existing registry tests still pass (no behavior change for `registry<T>` / `registry<T, Tag>` call sites)
- [x] New `KeyPayloadRegistryTest` suite covers: `intern(key, payload)` basics, sequential ids on distinct keys, first-write-wins on duplicate keys, `find(key)`, `get()` returning `Payload`, serialize/deserialize round-trip with `Key != Payload` (incl. lookup rebuild), Tag discrimination across `<Key, Payload>` pools, `clear()` wipe
- [x] Build with `BUILD_TESTING=ON` and run `ctest` — both `RegistryTest.*` and `KeyPayloadRegistryTest.*` pass

## Summary by CodeRabbit

* **New Features**
  * Registry now supports independent key and payload types for flexible data modeling
  * `intern()` method extended to accept key/payload pairs with deduplication by key
  * Serialization/deserialization preserves key/payload associations across round-trips
  * Independent singleton registries achieve isolated ID sequences via Tag parameter

* **Tests**
  * Added comprehensive test coverage for key/payload registry behavior and serialization round-trips

[Review Change Stack](https://app.coderabbit.ai/change-stack/genogrove/genogrove/pull/401?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: include/genogrove/data_type/registry.hpp [hunk 1]
```diff
@@ -265,18 +336,26 @@ class registry {
...
+                Payload value = serializer<Payload>::read(is);
+                auto id = static_cast<id_type>(new_storage.size());
+                new_storage.push_back(std::move(value));
+                new_lookup.emplace(new_storage.back(), id);
+            } else {
+                Key k = serializer<Key>::read(is);
+                Payload p = serializer<Payload>::read(is);
+                auto id = static_cast<id_type>(new_storage.size());
+                new_storage.push_back(std::move(p));
+                new_lookup.emplace(std::move(k), id);
```
[the remark this answers is not in the document]
Comment (riasc) [reply include/genogrove/data_type/registry.hpp]: **Verdict: Address.** Real correctness issue — and the new `Key != Payload` branch makes the failure mode worse than it was historically.

Trace: if a stream contains two entries with the same `Key`, the second `new_lookup.emplace(...)` returns `{it, false}` and is silently a no-op, while `new_storage` still grew. End state: `new_storage.size() > new_lookup.size()`. On the next `serialize()` in the `Key != Payload` branch, `key_by_id[i]` is `nullptr` for the orphaned slot and we dereference it — UB / null deref.

The `Key == Payload` branch has the same latent bug (pre-dates this PR), but the consequences are quieter because the historical serialize path walks `storage` directly and never reaches a null. Worth fixing in both branches for symmetry while we're here — check the `emplace` bool and throw `std::runtime_error("Failed to deserialize registry: duplicate key")`, matching the existing strong-exception-guarantee idiom in the same function. Will also add a regression test that builds a stream with a duplicate key and asserts the throw.

See the summary comment for the full review.

**Comment by riasc** [comment]
## Code Review

CI is green on all 7 build matrix jobs (gcc 13/14, clang 16/17/18, Apple clang 15/16). The change conforms to project conventions: C++20 concepts + `requires`, `std::runtime_error` for serialization failures, `[[nodiscard]]` on non-trivial returns, no trailing-underscore members, header-only template, Meyer's singleton + mutex pattern, strong-exception-guarantee deserialize idiom preserved.

### Bug
- **`deserialize()` does not reject duplicate keys** (`include/genogrove/data_type/registry.hpp:347-357`). Both branches `new_lookup.emplace(...)` and discard the `bool inserted` result. If a stream carries two entries with the same `Key`, `new_storage` grows but `new_lookup` does not, leaving `new_storage.size() > new_lookup.size()`. In the `Key != Payload` branch this surfaces on the next `serialize()`: the `key_by_id` index has `nullptr` at the orphaned slot and we dereference it — UB / null deref. The `Key == Payload` branch has the same latent bug (pre-dates this PR), but its consequences are quieter because the old serialize path walked `storage` directly. Fix: check the `emplace` bool in both branches and throw `std::runtime_error("Failed to deserialize registry: duplicate key")` to keep the strong-exception-guarantee contract. Detail and trace in the inline reply above.

### Suggestion (optional)
- Add a regression test that constructs a stream with two entries sharing a key and asserts `deserialize` throws (similar shape to the existing `DeserializeProvidesStrongExceptionGuaranteeOnTruncatedStream` and `DeserializeRejectsCountExceedingIdCapacity` tests). Goes hand-in-hand with the fix above.

### Notes
- Pre-existing CodeRabbit comment is the same finding — addressed in the inline reply.
- Template-param ordering decision (`<Key, Tag, Payload>` vs the issue's `<Key, Payload, Tag>`) is documented in the PR body and preserves the existing public surface; no concern.
- Wire format compatibility: historical `count + payloads` layout preserved when `Key == Payload`, so existing on-disk data still loads. New `count + (key, payload)` layout used only when types differ — clean separation, `if constexpr` switch, no runtime cost.

**Comment by riasc** [comment]
## Code Review (round 2 — after fix at 5a63f69)

CI is green on all 7 build matrix jobs at \`5a63f69\`. CodeRabbit acknowledged the fix and agreed the strong-exception-guarantee idiom is correct. Re-reviewing the fix commit and current branch state:

### Findings

**Lint** — \`tests/data_type/registry_test.cpp\` is missing a trailing newline at EOF. The new \`KeyPayloadRegistryTest.DeserializeRejectsDuplicateKeys\` block sits at the very end of the file and the append did not add a final \`\n\`. Trivial fix.

### Suggestions (optional)

**Symmetrize push/emplace order between deserialize branches.** CodeRabbit's follow-up flagged this:

\`\`\`cpp
if constexpr (key_is_payload) {
    new_storage.push_back(std::move(value));
    if (!new_lookup.emplace(new_storage.back(), id).second) throw …;   // push, then emplace
} else {
    if (!new_lookup.emplace(std::move(k), id).second) throw …;          // emplace, then push
    new_storage.push_back(std::move(p));
}
\`\`\`

Both orderings are functionally identical (same 1 copy + 1 move either way, since the unordered_map owns its own key copy). The asymmetry is a readability nit, not a correctness issue — the \`key_is_payload\` branch *has* to push first if it wants to use \`new_storage.back()\` as the emplace key. Swapping it would mean either using \`value\` directly (one extra copy on the happy path) or restructuring more. Not worth a commit unless we care about reader symmetry.

**CHANGELOG mention of the duplicate-key check.** The entry under \`[Unreleased]\` covers the new \`Payload\` parameter but not the duplicate-key rejection. Strictly, the fix closes a latent bug in the historical \`Key == Payload\` branch too (silent corruption → throw on malformed streams). Defensible to omit since legitimate \`serialize()\` never emits duplicates, but a one-line note inside the existing entry would be more honest.

### Re-verified

- Fix correctness in both \`if constexpr\` branches: trace ✓, strong exception guarantee preserved by local-temporaries-then-commit pattern.
- Both regression tests (\`RegistryTest.DeserializeRejectsDuplicateKeys\` and \`KeyPayloadRegistryTest.DeserializeRejectsDuplicateKeys\`) construct bad streams by hand, assert throw, and verify pre-existing singleton survives.
- \`gene_info\` works as an aggregate in C++20 despite having non-special member functions (operator==, serialize/static deserialize).
- No new inline comments on the fix commit; CodeRabbit's only response was the acknowledgement reply on the existing thread.
